### ISSUE TITLE
AWS Console login SessionDuration

### ISSUE THREAD

**Issue by pda** [99designs/aws-vault#77]
I've been trying to add `--session-duration` flag to `aws-vault login`, but hit a snag where adding the `SessionDuration` parameter to the federated login causes the following error:

> message Invalid credentials parameter
> description The request sent by the client was syntactically incorrect.

It's exactly as described a few days ago by somebody else, so perhaps a recent problem with AWS API:
https://forums.aws.amazon.com/thread.jspa?threadID=236788

The documentation at http://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_enable-console-custom-url.html suggests it should be possible to to set SessionDuration to up to 12 hours, regardless of the duration of the assumed role token being used to get the login token.

Below is my WIP, run out of time, and not ready to PR it without seeing it working against the API. Note also I based it on the older v3.4.0 config structure because `master` @ 69bf23b921017cde9f147d6e1241d3664ea666a1 is broken by https://github.com/99designs/aws-vault/pull/74

```diff
diff --git a/login.go b/login.go
index 033210b..57e239a 100644
--- a/login.go
+++ b/login.go
@@ -7,6 +7,8 @@ import (
 	"log"
 	"net/http"
 	"net/url"
+	"strconv"
+	"time"
 
 	"github.com/99designs/aws-vault/keyring"
 	"github.com/99designs/aws-vault/prompt"
@@ -16,11 +18,12 @@ import (
 )
 
 type LoginCommandInput struct {
-	Profile   string
-	Keyring   keyring.Keyring
-	MfaToken  string
-	MfaPrompt prompt.PromptFunc
-	UseStdout bool
+	Profile         string
+	Keyring         keyring.Keyring
+	MfaToken        string
+	MfaPrompt       prompt.PromptFunc
+	UseStdout       bool
+	SessionDuration time.Duration
 }
 
 func LoginCommand(ui Ui, input LoginCommandInput) {
@@ -59,14 +62,23 @@ func LoginCommand(ui Ui, input LoginCommandInput) {
 
 	q := req.URL.Query()
 	q.Add("Action", "getSigninToken")
+	if input.SessionDuration != 0 {
+		q.Add("SessionDuration", strconv.Itoa(int(input.SessionDuration.Seconds())))
+	}
 	q.Add("Session", string(jsonBytes))
 	req.URL.RawQuery = q.Encode()
 
+	ui.Debug.Printf("GET %s\n", req.URL)
+
 	resp, err := http.DefaultClient.Do(req)
 	if err != nil {
 		ui.Error.Fatal(err)
 	}
 
+	if contentType := resp.Header.Get("Content-Type"); contentType != "application/json" {
+		ui.Error.Println(fmt.Sprintf("expected application/json, got '%s'", contentType))
+	}
+
 	defer resp.Body.Close()
 	body, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
@@ -76,6 +88,7 @@ func LoginCommand(ui Ui, input LoginCommandInput) {
 	var respParsed map[string]string
 
 	if err = json.Unmarshal([]byte(body), &respParsed); err != nil {
+		ui.Error.Println("error parsing JSON:", string(body))
 		ui.Error.Fatal(err)
 	}
 
diff --git a/main.go b/main.go
index 4b7faa5..c0c1100 100644
--- a/main.go
+++ b/main.go
@@ -59,6 +59,7 @@ func main() {
 		loginProfile     = login.Arg("profile", "Name of the profile").Required().String()
 		loginMfaToken    = login.Flag("mfa-token", "The mfa token to use").Short('t').String()
 		loginStdout      = login.Flag("stdout", "Print login URL to stdout instead of opening in default browser").Short('s').Bool()
+		loginDuration    = login.Flag("duration", "Duration that the credentials for the console session are valid").Duration()
 		server           = kingpin.Command("server", "Run an ec2 instance role server locally")
 	)
 
@@ -128,11 +129,12 @@ func main() {
 
 	case login.FullCommand():
 		LoginCommand(ui, LoginCommandInput{
-			Profile:   *loginProfile,
-			Keyring:   keyring,
-			MfaToken:  *loginMfaToken,
-			MfaPrompt: prompt.Method(*promptDriver),
-			UseStdout: *loginStdout,
+			Profile:         *loginProfile,
+			Keyring:         keyring,
+			MfaToken:        *loginMfaToken,
+			MfaPrompt:       prompt.Method(*promptDriver),
+			UseStdout:       *loginStdout,
+			SessionDuration: *loginDuration,
 		})
 
 	case server.FullCommand():
```

/cc @lox @mtibben

**Comment by lox** [comment]
Hmmmm. Spent a long time tracing this back to try and remember the flow of authentication. For the benefit of those also digging: 

Initially `GetSessionToken` is used to get a session, then if a role is specified, `AssumeRole` is called. 

This table shows the limitations of each: http://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_request.html

I always thought the fact that a SessionToken / AssumeRole combination could call the federation login token API in the first place was very odd. It's not documented anywhere.

**Comment by lox** [comment]
When you are testing @pda, were you using a profile with a `role_arn` or without? I seem to recall `login` doesn't work at all if you use one without a `role_arn`.

**Comment by lox** [comment]
Which makes sense, because you need `AssumeRole` to call the API based on your initial link.

**Comment by lox** [comment]
In summary, looks to me like this should work. Feel like opening a ticket with AWS @pda? The IAM team will possibly know.

**Comment by pda** [comment]
I was going to watch the outcome of https://forums.aws.amazon.com/thread.jspa?threadID=236788 first. Reply from AWS so far:

> Are you using a set of IAM user credentials to call AssumeRole and then use the temporary credentials to request the sign-in token?

And yeah, I was using a `role_arn` profile which was working fine unless there was a `SessionDuration` in the request. I'm pretty sure it should work and there's a problem on their end. Next time I have a moment to look at it, I'll reproduce it again and report to AWS, assuming nothing comes of that forum thread.

**Comment by lox** [comment]
Looks like the ticket is updated.

**Comment by pda** [comment]
So Jeff from AWS says this should work:

> Long term credential (IAM User keys or SAML authentcation response) -> AssumeRole -> Use temporary credentials from AssumeRole to request a Sign-In Token -> Generate Sign-In URL

and this will not work:

> Long term credential (IAM User keys or SAML authentcation response) -> AssumeRole(1) -> Use temporary credentials from AssumeRole(1) to AssumeRole(2) -> Use temporary credentials from AssumeRole(2) to request a Sign-In Token -> Generate Sign-In URL

I would have thought we were doing the former, and not going two or more layers deep into AssumeRole temporary credentials before requesting a sign-in token and generating the URL. But I'd want to check that more carefully before jumping into that thread.

**Comment by lox** [comment]
Unfortunately not @pda. The whole thing that aws-vault does is keeps your long term credentials hidden by using `GetSessionToken` to create temporary credentials. Those credentials are then used to `AssumeRole` and Generate Sign-In URLs.

**Comment by pda** [comment]
Ah, so to adapt the non-working example from the JeffW@AWS:

> Long term credential (IAM User keys) -> GetSessionToken (1st layer) -> Use temporary credentials from GetSessionToken to AssumeRole (2nd layer) -> Use temporary credentials from AssumeRole to request a Sign-In Token -> Generate Sign-In URL

i.e. our first layer is GetSessionToken not AssumeRole, but presumably it's the same effect.

**Comment by pda** [comment]
Is there any reason for aws-vault to be requesting AssumeRole using temporary GetSessionToken credentials? Why not just AssumeRole using the long-term credentials and expose the temporary AssumeRole credentials? (I'm basing this entirely on the wording from the AWS rep, not an actual understanding of these mechanisms).

**Comment by bradfeehan** [comment]
> When you are testing @pda, were you using a profile with a role_arn or without? I seem to recall login doesn't work at all if you use one without a role_arn.

And you can't use a `role_arn` with IAM credentials from the same IAM account, right? So you can only use awe-vault to log in to a different account's console using a role?

(bit of an aside, possibly, sorry for derailing)

**Comment by lox** [comment]
Yeah, exactly @bradfeehan. I think that's why I did it the way it's implemented. For instance try using login with the core `99designs` bastion account @pda.

**Comment by bradfeehan** [comment]
Is that by design? I'd like to be able to log in to the main account, not sure what implications that has though. I could just set a username/password and log in normally so it's not a priority, was just curious.

**Comment by lox** [comment]
Not by design, simply that I couldn't make it work.

**Comment by lox** [comment]
Closed in #92
